jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c

```diff
@@ -2472,7 +2476,7 @@ ecma_builtin_array_prototype_object_reduce (ecma_value_t this_arg, /**< this arg
         bool k_present = false;
 
         /* 8.b */
-        while (!k_present && index < len && ecma_is_value_empty (ret_value))
+        while (!k_present && items_visited < len && ecma_is_value_empty (ret_value))
         {
           /* 8.b.i */
           ecma_string_t *index_str_p = ecma_new_ecma_string_from_uint32 (index);
```

Comment:
And here comes the catch: depending on `start_from_left` we read `index` or `len - 1 - index`. We can create a local variable for len - 1.

Reply:
This way no need int64 magic, no need to handle negative numbers.